java-checks/src/main/java/org/sonar/java/closeresource/ExecutionState.java

```diff
@@ -0,0 +1,269 @@
+      IdentifierTree identifier;
+      if (expression.is(Tree.Kind.MEMBER_SELECT)) {
+        identifier = ((MemberSelectExpressionTree) expression).identifier();
+      } else {
+        identifier = (IdentifierTree) expression;
+      }
+      if (isIgnoredCloseable(identifier.symbol())) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private boolean isIgnoredCloseable(Symbol symbol) {
+    if (CloseableVisitor.isCloseableOrAutoCloseableSubtype(symbol.type()) && !symbol.owner().isMethodSymbol()) {
+      return true;
+    } else {
+      CloseableOccurence currentOccurence = getCloseableOccurence(symbol);
+      return currentOccurence != null && currentOccurence.state.isIgnored();
+    }
+  }
+
+  void checkUsageOfClosables(List<ExpressionTree> expressions) {
+    for (ExpressionTree expression : expressions) {
+      checkUsageOfClosables(expression);
+    }
+  }
+
+  void checkUsageOfClosables(@Nullable ExpressionTree expression) {
+    if (expression != null) {
+      if (expression.is(Tree.Kind.IDENTIFIER) && CloseableVisitor.isCloseableOrAutoCloseableSubtype(expression.symbolType())) {
+        markAsIgnored(((IdentifierTree) expression).symbol());
+      } else if (expression.is(Tree.Kind.MEMBER_SELECT)) {
+        checkUsageOfClosables(((MemberSelectExpressionTree) expression).identifier());
+      } else if (expression.is(Tree.Kind.TYPE_CAST)) {
+        checkUsageOfClosables(((TypeCastTree) expression).expression());
+      } else if (expression.is(Tree.Kind.METHOD_INVOCATION)) {
+        checkUsageOfClosables(((MethodInvocationTree) expression).arguments());
+      } else if (expression.is(Tree.Kind.NEW_CLASS)) {
+        checkUsageOfClosables(((NewClassTree) expression).arguments());
+      }
+    }
+  }
+
+  void markAsIgnored(Symbol symbol) {
+    markAs(symbol, State.IGNORED);
+  }
+
+  void markAsClosed(Symbol symbol) {
+    markAs(symbol, State.CLOSED);
+  }
+
+  private void markAs(Symbol symbol, State state) {
+    if (closeableOccurenceBySymbol.containsKey(symbol)) {
+      closeableOccurenceBySymbol.get(symbol).state = state;
+    } else if (parent != null) {
+      CloseableOccurence occurence = getCloseableOccurence(symbol);
+      if (occurence != null) {
+        occurence.state = state;
+        closeableOccurenceBySymbol.put(symbol, occurence);
+      }
+    }
+  }
+
+  private Set<Tree> getUnclosedClosables() {
+    Set<Tree> results = Sets.newHashSet();
+    for (CloseableOccurence occurence : closeableOccurenceBySymbol.values()) {
+      if (occurence.state.isOpen()) {
+        results.add(occurence.lastAssignment);
+      }
+    }
+    return results;
+  }
+
+  @CheckForNull
+  private CloseableOccurence getCloseableOccurence(Symbol symbol) {
+    CloseableOccurence occurence = closeableOccurenceBySymbol.get(symbol);
+    if (occurence != null) {
+      return new CloseableOccurence(occurence.lastAssignment, occurence.state);
```

Comment:
:new: ![MAJOR](https://raw.githubusercontent.com/henryju/image-hosting/master/major.png)Remove this instantiation of "CloseableOccurence".[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2440)